File: src/android/iron_source/src/main/java/com/ee/internal/IronSourceBannerHelper.kt
package com.ee.internal

import android.graphics.Point
import androidx.annotation.AnyThread
import com.ee.Platform
import com.ironsource.mediationsdk.ISBannerSize

class IronSourceBannerHelper {
    companion object {
        @AnyThread
        private fun convertAdSizeToSize(adSize: ISBannerSize): Point {
            val density = Platform.getDensity()
            val width = adSize.width * density
            val height = adSize.height * density
            return Point(width.toInt(), height.toInt())
        }
    }

    private val _indexToSize: MutableMap<Int, Point> = HashMap()

    init {
        for (index in 0..3) {
            val adSize = getAdSize(index)
            val size = convertAdSizeToSize(adSize)
            _indexToSize[index] = size
        }
    }

    @AnyThread
    fun getAdSize(index: Int): ISBannerSize {
        if (index == 0) {
            return ISBannerSize.BANNER
        }
        if (index == 1) {
            return ISBannerSize.LARGE
        }
        if (index == 2) {
            return ISBannerSize.RECTANGLE
        }
        throw IllegalArgumentException("Invalid ad index")
    }

    @AnyThread
    private fun getIndex(adSize: ISBannerSize): Int {
        if (adSize == ISBannerSize.BANNER) {
            return 0
        }
        if (adSize == ISBannerSize.LARGE) {
            return 1
        }
        if (adSize == ISBannerSize.RECTANGLE) {
            return 2
        }
        throw IllegalArgumentException("Invalid ad size")
    }

    @AnyThread
    fun getSize(index: Int): Point {
        return _indexToSize[index] ?: throw IllegalArgumentException("Invalid ad index")
    }

    @AnyThread
    fun getSize(adSize: ISBannerSize): Point {
        return getSize(getIndex(adSize))
    }
}

File: src/ios/ee/iron_source/IronSourceBannerHelper.swift
//
//  IronSourceBannerHelper.swift
//  ee-x-d542b565
//
//  Created by eps on 1/25/21.
//

internal class IronSourceBannerHelper {
    private class func convertAdSizeToSize(_ adSize: ISBannerSize) -> CGSize {
        return CGSize(width: Double(Utils.convertDpToPixels(Float(adSize.width))),
                      height: Double(Utils.convertDpToPixels(Float(adSize.height))))
    }

    private var _indexToSize: [Int: CGSize] = [:]

    init() {
        for index in 0...3 {
            let adSize = getAdSize(index)
            let size = IronSourceBannerHelper.convertAdSizeToSize(adSize)
            _indexToSize[index] = size
        }
    }

    func getAdSize(_ index: Int) -> ISBannerSize {
        if index == 0 {
            return ISBannerSize(description: kSizeBanner, width: 320, height: 50)
        }
        if index == 1 {
            return ISBannerSize(description: kSizeLarge, width: 320, height: 90)
        }
        if index == 2 {
            return ISBannerSize(description: kSizeRectangle, width: 300, height: 250)
        }
        assert(false, "Invalid ad index")
        return ISBannerSize(description: kSizeBanner, width: 320, height: 50)
    }

    private func getIndex(_ adSize: ISBannerSize) -> Int {
        if adSize.sizeDescription == kSizeBanner {
            return 0
        }
        if adSize.sizeDescription == kSizeLarge {
            return 1
        }
        if adSize.sizeDescription == kSizeRectangle {
            return 2
        }
        assert(false, "Invalid ad size")
        return 0
    }

    private func getSize(index: Int) -> CGSize {
        guard let size = _indexToSize[index] else {
            assert(false, "Invalid ad index")
            return CGSize.zero
        }
        return size
    }

    func getSize(adSize: ISBannerSize) -> CGSize {
        return getSize(index: getIndex(adSize))
    }
}
